Optimize Venues search using React Concurrent Mode APIs

taneliang opened the pull request:
<!--
Thank you for contributing to NUSMods!
The template below was made to help both you and the reviewers understand
your changes. Please fill it up to the best of your ability.
(These are comments, they won't be shown in the "preview")
-->

### TL;DR
<!-- Please link to a Github issue (type `#` to autocomplete issue) -->
<!-- Or provide a brief explanation about the problem -->

Uses React Concurrent Mode APIs (added in #3039) to optimize the Venues components reworked in #3038.

### Optimizations in this PR

1. Uses `useDeferredValue` to allow the search box to be immediately updated in a sync render before a second, longer concurrent render is performed.

	|`master`|This PR|
	|-|-|
	|![boxm](https://user-images.githubusercontent.com/12784593/101497229-b68ccd00-39a5-11eb-951f-aee9166efea0.gif)|![box](https://user-images.githubusercontent.com/12784593/101497067-847b6b00-39a5-11eb-9800-8de9fcdec7a8.gif)|

	This may sound similar to the DIY deferred update using 2 nested `requestAnimationFrame` calls that we currently have on `master`. However, that is inferior to this solution because:
	1. CM allows in-progress renders to be abandoned, allowing multiple keystrokes in quick-ish succession to remain responsive, as previous in-progress renders are simply abandoned. On `master`, the app noticeably freezes up once a legacy render starts as it cannot be interrupted. In the gifs above, notice that on `master`, a bunch of characters appear at once -- this happens because I typed while the app was frozen.
	1. (Independent of CM) #3038 reduces unnecessary renders of expensive components, which reduces the amount of time the browser spends frozen.

	This is what typing a single character looks like in the [React Scheduling Profiler](https://react-devtools-scheduling-profiler.vercel.app/) (that I helped create :D):

	![image](https://user-images.githubusercontent.com/12784593/101497659-2d29ca80-39a6-11eb-961d-1d1f8acc4c69.png)

	TL;DR of how to read the React data (the stuff above the grey bar):
	- The circles represent state updates in this case.
	- Each horizontal row = a React fiber lane (each of which has a purpose/priority)
	- A blue box in each lane represents a render chunk, and a red box is a commit phase.

	Notice the short sync render on the left (which is in response to the key press), followed by a very long render caused by the deferred value updating. Also notice that `VenueList` takes a long time to render.

	This is what typing a second character during a render looks like:

	![image](https://user-images.githubusercontent.com/12784593/101498958-a70e8380-39a7-11eb-9ed0-eb1206b06763.png)
	
	Notice that a concurrent render got abandoned.

1. Uses `useDeferredValue` to allow the availability filter to open instantly. On `master`, this just freezes as there were no optimizations for this.
	![ffr](https://user-images.githubusercontent.com/12784593/101496627-f30bf900-39a4-11eb-8ab5-d214e7d35549.gif)

Note that we do not also defer `searchOptions`, as I couldn't get it to feel right. It seems better to just let the page freeze while the render happens. (Edit: it's now deferred. See GIFs in this comment below: https://github.com/nusmodifications/nusmods/pull/3040#issuecomment-748069917)

taneliang committed:
Improve venue search box responsiveness

taneliang committed:
useDeferredValue(isAvailabilityEnabled) with spinner to reduce page freezing

taneliang commented on website/src/views/venues/VenuesContainer.tsx:
```diff
@@ -153,23 +167,25 @@ export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
           type="button"
         >
-          <Clock className="svg" /> Find free rooms
+          {isAvailabilityEnabled !== deferredIsAvailabilityEnabled ? (
+            <LoadingSpinner className={styles.availabilitySpinner} white={isAvailabilityEnabled} />
```
This causes the spinner to flash on faster browsers. No idea how to fix it, but I think storing `isAvailabilityEnabled !== deferredIsAvailabilityEnabled` in a `useState` and using our `defer` may be able to fix it at the cause of more renders?

chrisgzf replied:
How do I repro this flashing? I tried to profile it but I couldn't see the flashing

![image](https://user-images.githubusercontent.com/4933577/101523869-227e2e00-39c4-11eb-862d-d8a3554e2698.png)

taneliang replied:
Oh, just click the Find free rooms button. The clock icon in the button will disappear for a moment as the spinner flashes

ZhangYiJiang replied:
I'm not sure if putting the spinner here is the best idea - the thing that's "loading" is the venue list below, and it's quite easily missed (though the animation helps). Won't putting a spinner there instead work? 

Also thinking about https://reactjs.org/docs/concurrent-mode-patterns.html#delaying-a-pending-indicator, but I don't know how to make that work with useDeferredValue

taneliang replied:
Ooh, deferring the loading indicator in CSS is brilliant. I'll switch us to `useTransition` then, especially since the current condition for determining whether to show the indicator or not is a bit of a hack.

taneliang replied:
I lied, `useTransition` isn't a good fit for our use case. It's back to `useDeferredValue`, but with a deferred spinner :D

taneliang commented on website/src/views/venues/VenuesContainer.tsx:
```diff
@@ -153,23 +167,25 @@ export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
         <button
           className={classnames(
             'btn btn-block btn-svg',
-            styles.availabilityToggle,
```
This style didn't exist

taneliang commented on website/src/views/venues/VenuesContainer.scss:
```diff
@@ -113,6 +113,17 @@ $venue-list-width: 16rem;
   border-radius: 0 0 $border-radius $border-radius;
 }
 
+.availabilitySpinner {
+  display: inline-block;
+  width: 1.4rem;
+  height: 1.4rem;
+  margin: 0 0.3rem -0.1rem 0;
+  border-width: 0.2rem;
+  // Use parent button's text color so that the spinner is still visible when
+  // the button is hovered over.
+  border-left-color: initial;
+}
```
This is copied and customized from `LoadingSpinner.scss`. Not particularly clean, but I couldn't find any other way using CSS specificity to get CSS to use these styles over the `LoadingSpinner` ones, and I didn't have much time to work on this. Help appreciated!

ZhangYiJiang replied:
Won't nesting them under `.venueSearch` work?

taneliang replied:
Oh yes. Derped because I was rushing. Anyway this style has been removed, so I'm resolving this convo

taneliang commented:
Rebased on parent branch, and fixed conflict with [`d5b270e` (#3038)](https://github.com/nusmodifications/nusmods/pull/3038/commits/d5b270ed29d5b7ddeccc55388bca5d0915aba348)

taneliang committed:
Move spinner to an overlay

taneliang committed:
Reduce defer delay to improve perceived responsiveness

taneliang committed:
Defer searchOptions

taneliang committed:
Switch back to useDeferredValue as it's too messy to manage both primary and deferred values manually

taneliang commented:
Main changes:

- Spinner is now overlaid on `VenueList`
- Spinner and overlay are now deferred by 0.2s. The experience is far better.
- `searchOptions` is now deferred -- selecting stuff in the Find free rooms is now super smooth 😍 

Updated gifs:

|No throttling|6x CPU throttling|
|-|-|
|![defno](https://user-images.githubusercontent.com/12784593/102616865-361f5680-4173-11eb-9ab7-2af82f88a9a8.gif)|![def6](https://user-images.githubusercontent.com/12784593/102616869-3881b080-4173-11eb-9032-5ada38ae106c.gif)|

taneliang commented:
Rebased on parent branch, no diff change

ZhangYiJiang commented on website/src/styles/utils/deferred.scss:
```diff
@@ -0,0 +1,11 @@
+// Source: https://reactjs.org/docs/concurrent-mode-patterns.html#delaying-a-pending-indicator
+.deferred {
+  visibility: hidden;
+  animation: 0s linear 0.2s forwards makeVisible;
```
We could in theory make this configurable from the component side using CSS variables or even inline styles, but probably YAGNI

taneliang replied:
Yeah, probably not a good idea to overengineer this right now hahaa

ZhangYiJiang commented on website/src/views/components/SearchBox.tsx:
```diff
@@ -16,18 +16,18 @@ import styles from './SearchBox.scss';
   placeholder?: string;
   onChange: (value: string) => void;
-  onSearch: () => void;
+  onSearch?: () => void;
```
Might be useful adding a small comment on the difference between this and onChange

ZhangYiJiang commented on website/src/views/modules/ModuleFinderContainer/ModuleFinderContainer.tsx:
```diff
@@ -73,7 +74,7 @@ const ModuleFinderContainer: React.FC = () => (
             />
 
             <LoadingComponent>
-              <div className={styles.loadingOverlay} />
```
Should remove the unused duplicate class here from `ModuleFinderContainer.scss`

taneliang replied:
Yep, it was already removed

ZhangYiJiang commented on website/src/views/venues/VenuesContainer.tsx:
```diff
@@ -46,18 +54,16 @@ export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
-  ] = useState<string>(() => qs.parse(location.search).q || '');
-  /** Actual string to search with; deferred update */
-  const deferredSearchQuery = searchQuery; // TODO: Redundant now. Use React.useDeferredValue after we adopt concurrent mode
+  const [searchQuery, setSearchQuery] = useState<string>(() => qs.parse(location.search).q || '');
```
/nit 

```suggestion
  const [searchQuery, setSearchQuery] = useState(() => qs.parse(location.search).q || '');
```

taneliang replied:
That doesn't work because `qs.parse` returns `any`. We could do the below type assertion instead, but the current code seems good enough

```suggestion
  const [searchQuery, setSearchQuery] = useState(
    () => (qs.parse(location.search) as Params).q || '',
  );
```

ZhangYiJiang commented on website/src/views/venues/VenuesContainer.tsx:
```diff
@@ -46,18 +54,16 @@ export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
-  /** Actual string to search with; deferred update */
-  const deferredSearchQuery = searchQuery; // TODO: Redundant now. Use React.useDeferredValue after we adopt concurrent mode
+  const [searchQuery, setSearchQuery] = useState<string>(() => qs.parse(location.search).q || '');
+  const handleSearchChange = useCallback((newSearchQuery: string) => {
+    setSearchQuery(newSearchQuery);
+  }, []);
```
Isn't `handleSearchChange` exactly the same as `setSearchQuery`?

taneliang replied:
Oh yes, I think I forgot to change this back after my brief experiment with `useTransition`

ZhangYiJiang commented on website/src/views/venues/VenuesContainer.tsx:
```diff
@@ -213,11 +233,15 @@ export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
-        {size(matchedVenues) === 0 ? (
-          renderNoResult()
-        ) : (
+        <div className="position-relative">
```
I think without specifying height this will not take up the whole height of the venues list right?

taneliang replied:
I think it does since it wraps the venue list component

![image](https://user-images.githubusercontent.com/12784593/102706993-f6bb4c00-42d1-11eb-9970-42fb2acc8951.png)

ZhangYiJiang reviewed (commented):
I'm going to need a bit more time reading up the concurrent mode docs to understand what useDeferredValue is doing here

taneliang commented:
@ZhangYiJiang I think reading `useDeferredValue`'s implementation may be easiest. https://github.com/facebook/react/blob/99554dc36fa9f5ef29b75dd59836dcc720d831b9/packages/react-reconciler/src/ReactFiberHooks.old.js#L1498-L1538

The `mount*` functions are called the first time the hook is called, and the `update*` functions are called on subsequent calls. The `*State` and `*Effect` functions correspond to `useState` and `useEffect`.

TL;DR it looks like it's just storing the previous state in a `useState`, then doing a state update in a passive effect. But interestingly it also sets a `transition` flag on the fiber (which is now just `1` or `0`, but in the past it was the suspense config object that you see in the CM docs). I'm not sure yet what `transition` accomplishes, but `useTransition` does that as well.

taneliang committed:
Fix review comment: add `onSearch` and `onChange` docstrings

taneliang committed:
Fix review comment: remove unnecessary `handleSearchChange`

taneliang committed:
Make SearchBox throttle prop optional
